### ISSUE TITLE
Accessibility: Missing status messages

### ISSUE THREAD

**Issue by vikrampm1** [woothemes/woocommerce#51902]
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

### Issue Description:

After updating the product quantity of an item in the cart or removing an item from the cart, the page refreshes but there is no visual or auditory success message that the item was removed or the quanty was updated successfully. Blind users need to explore the page to find out if the item removal or quantity update worked.

### Component or Page: 

Shortcode: [woocommerce_cart] & Cart Classic Block

### Name affected element:

Updating cart or removing item from cart

### Relevant code snippets:

N/A

### Recommended fix:

Ideally, cart updates and product removals can happen with ajax rather than a page reload and a visual and auditory success message can be added above the product table. If a page refresh is required, add the success message and set keyboard focus on the success message to ensure that it is read out. This can be the same banner that exists on the block version of the cart page.

**Comment by mikejolley** [comment]
> Ideally, cart updates and product removals can happen with ajax rather than a page reload and a visual and auditory success message can be added above the product table. If a page refresh is required, add the success message and set keyboard focus on the success message to ensure that it is read out. This can be the same banner that exists on the block version of the cart page.

This sounds like a large refactor. I'm not sure if we should be doing something of this scope with the legacy shortcodes when the new experience takes care of this much better.

cc @opr what do you think?

**Comment by vikrampm1** [comment]
@opr any comment on the above?

**Comment by jeffpaul** [comment]
@mikejolley @opr my concern with considering NOT doing something here is what the known or expected breakdown in existing Woo storefronts using classic vs. block-based themes?  My gut says the classic theme approach is not likely a statistically insignificant portion and thus still a large vector for inaccessible stores on this topic.  Perhaps we could look at decomposing the work needed here to make resulting PR(s) smaller in scope, easier to review, and hopefully quicker to get into core and work towards improving the accessibility for these sorts of storefronts?

**Comment by opr** [comment]
> > Ideally, cart updates and product removals can happen with ajax rather than a page reload and a visual and auditory success message can be added above the product table. If a page refresh is required, add the success message and set keyboard focus on the success message to ensure that it is read out. This can be the same banner that exists on the block version of the cart page.

> This sounds like a large refactor. I'm not sure if we should be doing something of this scope with the legacy shortcodes when the new experience takes care of this much better.

I agree, changing to use AJAX seems like a large refactor, however the second suggestion of adding a message seems OK. 

> My gut says the classic theme approach is not likely a statistically insignificant portion and thus still a large vector for inaccessible stores on this topic. 

I suspect you're right, we should still consider making sure the shortcode checkout is accessible.

> Perhaps we could look at decomposing the work needed here to make resulting PR(s) smaller in scope, easier to review, and hopefully quicker to get into core and work towards improving the accessibility for these sorts of storefronts?

If you would go with the AJAX route, yeah I agree, however the alternative suggested seems OK?

**Comment by jeffpaul** [comment]
@Manussakis any concern on the alternative route to resolve here or would that make sense to you?

**Comment by Manussakis** [comment]
@mikejolley @opr @jeffpaul  These are alternatives I came up with to avoid implementing AJAX:

### Classic Cart
**Update quantity**
The quantity input is inside the context of the product row in the cart table, so users know what it’s related to. Quantity updates are announced by screen readers and there is a visual indication of that, so I don't believe any work in this section is needed.
![Image](https://github.com/user-attachments/assets/b6408cd4-8c7e-4c55-b7fa-df1d7c6adf92)


**Remove item**
When a product is removed, new HTML content comes from the server as a response, and then it’s inserted into the page via JS, so there is no page reload. That's why the [focus_populate_live_region](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/js/frontend/woocommerce.js#L198) function is not called and the alert message is not announced. The solution can be just to focus on the alert message already in place via JS.
![Image](https://github.com/user-attachments/assets/f2e3ae0d-2bf7-4849-95fc-14fef788b83b)


### Block-based Cart
**Update quantity**
The live region that announces the change in the quantity has its content replaced by a shipping information message. e.g. "1 shipping option was found." This prevents the quantity message from being completely read. A fix can be creating a separate live region dedicated to the quantity updates or avoiding its content being replaced by the shipping message too soon.

**Remove item**
Screen readers already announce the product removal. This is done by calling the [speak function](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/cart-line-items-table/cart-line-item-row.tsx#L331). What is missing is a visual indication of the change which can be done using the snackbar block. The same solution is used on the block-based Checkout page when a coupon is added/removed.
![Image](https://github.com/user-attachments/assets/0fd6cc62-ca1c-4426-bc67-fc17066fc700)

**Comment by mikejolley** [comment]
> The live region that announces the change in the quantity has its content replaced by a shipping information message. e.g. "1 shipping option was found." This prevents the quantity message from being completely read. A fix can be creating a separate live region dedicated to the quantity updates or avoiding its content being replaced by the shipping message too soon.

Could the shipping one be somehow prevented on update of quantity? I would expect shipping to only notify when using the shipping calculator.

**Comment by Manussakis** [comment]
@mikejolley 

> Could the shipping one be somehow prevented on update of quantity? I would expect shipping to only notify when using the shipping calculator.

It makes sense, I will investigate that.
As for my other suggestions, can I go ahead and implement them?

**Comment by mikejolley** [comment]
I think the other points sound ok 👍🏻

**Comment by jeffpaul** [comment]
@Manussakis you're no longer blocked on this, correct?

**Comment by Manussakis** [comment]
@amberhinds
This issue was flagged as "Not Fixed" in the 2024-09 WooCommerce Core Audit spreadsheet.

I followed the steps to QA in the PR description, and it worked as expected.

### Remove item from cart
**Classic Cart**
https://github.com/user-attachments/assets/58fc99b1-633b-42ba-b92f-fbfa00e2fe9a

**Block-based Cart**
https://github.com/user-attachments/assets/b95aef55-5f84-4752-b004-d6e70c34259f

### Update item quantity
**Classic Cart**
No changes have been made, please check the "Classic Cart -> Update quantity" section of [this comment](https://github.com/woocommerce/woocommerce/issues/51902#issuecomment-2489098406) for more context: 

**Block-based Cart**
https://github.com/user-attachments/assets/f3ddc231-ac4d-4c2c-af27-fc0ff26f84d5

Could you please double-check it?
Thanks!

cc: @vikrampm1

**Comment by amberhinds** [comment]
This is the page we're testing the shortcode on: https://woo.equalizedigital.com/cart-shortcode/ (password: wooa11y) it's running a dev version of 9.9.

Here's what I hear when updating the product quantity or removing items from the cart:

https://github.com/user-attachments/assets/22f3ccf5-a433-4259-b30e-a10da0fd1e29

@mikejolley said in the linked mcomment:
>Quantity updates are announced by screen readers and there is a visual indication of that.

It is true that when the quantity field is changed, there is an auditory announcement, but changing the quantity field does not save the change. There is an "Update Cart" button that is added below the table, which must be clicked to save changes and truly change the quantity. This is the action that does not have a success message confirming that the quantity is changed.

I'm not able to get the removed message to appear visually on this page as shown in your screenshot. So, removing items from the cart is also not announcing anything and does not have a visual success message for sighted people.

All of the videos in the linked PR appear to show the block-based cart, not the shortcode version this issue references, and that I have shown in my video.

I know the shortcode is not the preferred way for people to display the cart, but many stores, especially those built will page builders likely still utilize it and will continue to do so for a long time in the future.

**Comment by Manussakis** [comment]
@amberhinds 
I was not able to use this page to test the Cart: https://woo.equalizedigital.com/cart-shortcode/.
As much as I add products to the Cart, the page remains empty:

![Image](https://github.com/user-attachments/assets/c6ad6850-a2bc-426c-accd-c25d5a7e0442)

But I can see the products in the Cart if I go to the Cart page (https://woo.equalizedigital.com/cart/)

![Image](https://github.com/user-attachments/assets/7fe84b05-9bbc-4210-87b9-e88df8f46d13)

Please make sure the Cart page you are testing is the same as the one set in the WooCommerce -> Settings -> Advanced page in the admin:

![Image](https://github.com/user-attachments/assets/69b1fc21-1870-4880-8aa7-2261caccefc4)

Please check my test on the Cart Shortcode page:
**Update quantity**
![Image](https://github.com/user-attachments/assets/429c0a5a-49c1-4806-8a2d-faeef2e5527b)

**Item removal**
![Image](https://github.com/user-attachments/assets/9db7857b-2962-4062-a2f5-cf77e9e2657d)

**The full video** 

https://github.com/user-attachments/assets/ed9ac448-f188-48f7-9248-183dc8895e72

---

>  There is an "Update Cart" button that is added below the table, which must be clicked to save changes and truly change the quantity. 

You are right, clicking on this "Update Cart" button shows the success message at the top, but it's not announced by screen readers.

![Image](https://github.com/user-attachments/assets/36ff0be8-deee-41bc-ae6c-afa1e86f44ac)

I suggest creating a new issue to announce the "cart updated" message once the "Update Cart" button is clicked.

cc: @vikrampm1

**Comment by amberhinds** [comment]
I have created a new issue as requested. https://github.com/woocommerce/woocommerce/issues/58121
